doc/vxlan/Overlay ECMP with BFD.md

```diff
@@ -0,0 +1,216 @@
+- Provide support for endpoint modification for a route prefix. Require SAI support for SET operation of routes. 
+- Provide support for endpoint deletion for a route prefix. Orchagent shall check the existing entries and delete any tunnel/nexthop based on the new route update
+- Ensure backward compatibility with single endpoint routes
+- Use SAI_NEXT_HOP_GROUP_MEMBER_ATTR_WEIGHT for specifying weights to nexthop member
+- Desirable to have per tunnel stats via sai_tunnel_stat_t 
+ 
+#### Detailed flow
+
+VnetOrch is one of the critical module for supporting overlay ecmp. VnetOrch subscribes to VNET and ROUTE updates from APP_DB.
+
+When a new route update is processed by the add operation,  
+
+1. VnetOrch checks the nexthop group and if it exists, reuse the group 
+2. For a new nexthop group member, add the ECMP member and identify the corresponding monitoring IP address. Create a mapping between the monitoring IP and nexthop tunnel endpoint. 
+3. Initiate a BFD session for the monitoring IP if it does not exist 
+4. Based on the BFD implementation (BfdOrch vs Control plane BFD), subscribe to BFD state change, either directly as subject observer (similar to port oper state notifications in orchagent) or via STATEDB update. 
+5. Based on the VNET global configuration to advertise prefixes, indicate to STATEDB if the prefix must be advertised by BGP/FRR only if there is atleast one active nexthop. Remove this entry if there are no active nexthops indicated by BFD session down so that the network pfx is no longer advertised.  
+
+#### Monitoring Endpoint Mapping 
+
+VNET_ROUTE_TUNNEL_TABLE can provide monitoring endpoint IPs which can be different from the tunnel termination endpoints. VnetOrch creates a mapping for such endpoints and based on the monitoring endpoint (MonEP1) health, proceed with adding/removing nexthop tunnel endpoint (EP1) from the ECMP group for the respective prefix. It is assumed that for one tunnel termination endpoint (EP1), there shall be only one corresponding monitoring endpoint (MonEP1).  
+
+#### Pros of SWSS to handle route update based on tunnel nexthop health: 
+
+- No significant changes, if BFD session management is HW offload via SAI notifications or Control Plane assisted. 
+- Similar to NHFLAGS handling for existing route ECMP group 
+- Better performance in re-programming routes in ASIC instead of separate process to monitor and modify each route prefix by updating DB entries 
+
+### BfdOrch
+Sonic may offload the BFD session handling to hardware that has BFD capabilities.  A new module, BfdOrch shall be introduced to handle BFD session to monitoring endpoints and check the health of remote endpoints. BfdOrch shall offload the session initiation/sustenance to hardware via SAI APIs and gets the notifications of session state from SAI. The session state shall be updated in STATE_DB and to any other observer orchestration agents.  
```

Comment:
The path from SYNCD to ORCH is quite slow, in such cases the BFD state notification from SYNCD to BFDOrch may add a delta to the desired BFD timeout. The actual BFD timeout achieved may be significantly higher than desired timeout.